kotlin-extension-functions/build.gradle

```diff
@@ -0,0 +1,52 @@
+buildscript {
+    repositories {
+        mavenCentral()
+        jcenter()
+    }
+    dependencies {
+        classpath "org.jetbrains.dokka:dokka-gradle-plugin:${dokka_version}"
+    }
+}
+repositories {
+    jcenter() // or maven { url 'https://dl.bintray.com/kotlin/dokka' }
+}
+
+apply plugin: 'org.jetbrains.dokka'
+apply plugin:"kotlin-kapt"
+
+sourceCompatibility = 1.8
+compileKotlin { kotlinOptions.jvmTarget = "1.8" }
+compileTestKotlin { kotlinOptions.jvmTarget = "1.8" }
+
+dependencies {
+    annotationProcessor "io.micronaut:micronaut-inject-java:$micronautVersion"
+    kapt "io.micronaut:micronaut-inject-java:$micronautVersion"
+
+    implementation "com.fasterxml.jackson.module:jackson-module-kotlin:2.10.2", {
+        exclude group:"org.jetbrains.kotlin", module:'kotlin-reflect'
+    }
+    implementation "io.micronaut:micronaut-inject:$micronautVersion"
+    implementation "io.micronaut:micronaut-runtime:$micronautVersion"
+    implementation "io.micronaut:micronaut-http-server:$micronautVersion"
+    implementation "io.micronaut:micronaut-http-client:$micronautVersion"
```

Comment:
Adding this module will add all of these dependencies on the runtime classpath of applications that use them. I don't think we can safely assume users want runtime, http-server, or http-client. What can be done to only enable the extensions if the dependency is present?

Reply:
I'm not *super* familiar with all the various gradle lifecycle configurations for dependencies; is there a alternative to `implementation` (and something corresponding in source?) to mark classes to be optionally loaded only if something they depend on is present? Perhaps that's possible, I'm not sure, but I don't think I've heard of such a thing.

Alternatively, I *could* separate the extensions for those modules into corresponding kotlin modules (like having a `micronaut-runtime-kotlin-ext` subproject, a `micronaut-http-client-kotlin-ext` subproject, etc).

Reply:
> is there a alternative to implementation (and something corresponding in source?) to mark classes to be optionally loaded only if something they depend on is present? 

I think what you want is `compileOnly`. It will allow the code to compile but will not load these dependencies at runtime, allowing users to optionally include them on their project's dependencies. More info [here](https://docs.gradle.org/current/userguide/building_java_projects.html#sec:java_dependency_management_overview).

Reply:
@edrd-f Yes, however I don't know how Kotlin handles extension classes for classes that aren't on the classpath. If it simply ignores them then great.

Reply:
@jameskleeh did a test now and it behaves just like Java: `NoClassDefFoundError`, which makes sense since it's not possible to get an instance of the class prior to the extension call.